Update distributed tracing docs (#23313)
dotnet/docs: docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md

---
title: Collect a distributed trace - .NET

> description: A tutorial to collect distributed traces in .NET applications

Reviewer: We may want to add `OpenTelemetry` in the description to get some SEO boost.

ms.topic: tutorial
ms.date: 03/14/2021
---

# Collect a distributed trace

> **This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 

Reviewer 2: >✔️ [](start = 27, length = 2)

why you have this symbol?

Reviewer: All the diagnostics docs have this symbol for marking applicable .NET versions

> **This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 

Reviewer 3: ```suggestion
**This article applies to: ✔️** .NET 5.0 and later versions **or** any .NET application using the 
```

> **This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 

Reviewer 4: minor:
I'd try to word this to say that - This is applicable for all .NET versions, and then further clarify that - .NET 5 or higher, requires no extra dependency, older versions require installing SD.DS package.
This will avoid any potential misunderstanding that this is only a new thing and only works for newer .NET only.
(#1 question we received in otel as well!)

Author: Good feedback. I've update the 'applies to' guidance to indicate .NET Framework >= 4.5 and .Net Core >= 2.1. The need to include DiagnosticSource NuGet is embedded in the tutorial text but no longer mentioned here.

> [DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/5.0.1) version 5 or later

Reviewer 3: ```suggestion
[DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/) version 5 or later
```

> .NET applications can be instrumented using the <xref:System.Diagnostics.Activity?displayProperty=nameWithType> API to produce

Reviewer 3: Is the term `application` here referring to any application and library?

Author: Yes that was intent, but I changed 'application' -> 'code' to avoid any implication that this is only for executables.

> distributed tracing telemetry. In this tutorial you will record this instrumented telemetry with different telemetry collection

Reviewer 4: record this instrumented telemetry with different telemetry collection libraries -- this sentence is bit hard to follow.

Author: yep : ) Updated the wording.

libraries so that it is available to diagnose application issues. See 
[the instrumentation tutorial](distributed-tracing-instrumentation-walkthroughs.md) if you need to add new instrumentation.

## Collect using OpenTelemetry

### Prerequisites

> - [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet) or a later version

Reviewer 4: Why 3.1 as the pre-req? This would work even in 2.1 right?

Author: Probably, I just didn't test it. I'm going to leave it as-is for now but if customers care we could re-test the workflow using an older SDK to ensure I didn't accidentally use a 3.1 only feature.

### Create an example application

First you will create an example application that has some distributed trace instrumentation but nothing is being collected.

```dotnetcli
dotnet new console
```

Applications that target .NET 5 and later already have the necessary distributed tracing APIs included. For apps targeting older
.NET versions add the [System.Diagnostics.DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/)
version 5.0.1 or greater.
```dotnetcli

> dotnet add package System.Diagnostics.DiagnosticSource --version 5.0.1

Reviewer 5: ```suggestion
dotnet add package System.Diagnostics.DiagnosticSource
```

I'd rather not have versions burned into the doc since they tend to go stale

Author: For some reason I had concluded that was important, but retesting without it worked fine so let's say it was user error : )

```

Replace the contents of the generated Program.cs with this example source:
```C#
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Threading.Tasks;

> namespace Sample.DistributedTracing

Reviewer 2: >DistributedTracing [](start = 17, length = 18)

This example is good but not using the created activities at all which can be confusing to the reader why we are creating activity at all.

Author: Did you have a suggestion for what to change? The text at the top currently says "First you will create an example application that has some distributed trace instrumentation but nothing is being collected"

Reviewer 4: We can leave a code comment "In this example, the activity will not be created at all, as there is no registered listener."

Author: I updated the text above the code snippet to make it hopefully more explicit : )

{
    class Program
    {
        static ActivitySource s_source = new ActivitySource("Sample.DistributedTracing");

        static async Task Main(string[] args)
        {
            await DoSomeWork();
            Console.WriteLine("Example work done");
        }

        static async Task DoSomeWork()
        {
            using (Activity a = s_source.StartActivity("SomeWork"))
            {
                await StepOne();
                await StepTwo();
            }
        }

        static async Task StepOne()
        {

>             using (Activity a = s_source.StartActivity("StepOne"))

Reviewer 5: Do we want some explicit guidance on what meets the bar for Activity creation? Creating Activities is expensive and I don't this to be misunderstood as create an Activity for every method.

Author: There is a little bit of guidance in the instrumentation tutorial. Given this is the collection tutorial we'd be wandering a bit off topic to discuss best practices for instrumentation.

            {
                await Task.Delay(500);
            }
        }

        static async Task StepTwo()
        {
            using (Activity a = s_source.StartActivity("StepTwo"))
            {
                await Task.Delay(1000);
            }
        }
    }
}
```

Running the app does not log anything yet
```dotnetcli
> dotnet run
Example work done
```

> ### Add logging using OpenTelemetry

Reviewer 4: I'd avoid the work "logging" here as it could be confused with the "logging" pillar of Observability. Would prefer to consistently use `tracing` in this doc.

Author: I reworded as "Collect a trace using OpenTelemetry".

I am guessing that customers will relate better to 'trace' or 'tracing data' as nouns rather than 'trace/tracing' as a verb. I'll try to be consistent using the word 'collect' when I am drawing attention to the work of capturing data and transmitting it to a backend.

Add the [OpenTelemetry](https://www.nuget.org/packages/OpenTelemetry/) and 
[OpenTelemetry.Exporter.Console](https://www.nuget.org/packages/OpenTelemetry.Exporter.Console/) NuGet packages.

```dotnetcli
dotnet add package OpenTelemetry
dotnet add package OpenTelemetry.Exporter.Console
```

Update Program.cs with additional using statments:
```C#
using OpenTelemetry;
using OpenTelemetry.Trace;

> using System;

Reviewer 3: minor: put System before OpenTelemetry?

Author: By default the IDE sorts the usings in alphabetical order so OpenTelemetry goes first. I figured it was better to use the default ordering than to customize it.

using System.Diagnostics;
using System.Threading.Tasks;
```

And update Main() to create the OpenTelemetry TracerProvider:
```C#
        public static async Task Main()
        {
            using var tracerProvider = Sdk.CreateTracerProviderBuilder()

>                 .SetSampler(new AlwaysOnSampler())

Reviewer 5: ```suggestion
               .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MySample"))
                .SetSampler(new AlwaysOnSampler())
```

Reviewer 4: To keep this a minimal code, I'd suggest to remove both Sampler and Resource concept.

Author: I have removed both

Author: I've added Resource

                // Add more libraries

>                 .AddSource("Sample.DistributedTracing")

Reviewer 4: It'd be good to clarify that this name must match the name of AS created earlier.

Author: Updated the text after the example to clarify this

                // Add more exporters
                .AddConsoleExporter()
                .Build();

            await DoSomeWork();
            Console.WriteLine("Example work done");
        }
```

Now the app logs distributed trace information to the console:
```dotnetcli
> dotnet run
Activity.Id:          00-35c0e68b0dac3c49be08a9d9cab32579-0b7477e11aa20d40-01
Activity.ParentId:    00-35c0e68b0dac3c49be08a9d9cab32579-d95f666d24193f40-01
Activity.DisplayName: StepOne
Activity.Kind:        Internal
Activity.StartTime:   2021-03-15T01:58:10.7661575Z
Activity.Duration:    00:00:00.5013965
Resource associated with Activity:

>     service.name: unknown_service:temp

Reviewer 5: Update service name as mentioned in comment above

Author: @cijothomas recommend removing the API call so we have to decide, do we want simpler code or do we want this field populated.

I'm leaning toward including it in the example because it seems like something a typical OpenTelemetry user will need to do and understand. Thoughts?

Reviewer 4: If the example should look neater, then yes, need to add Resource in the code example.

Activity.Id:          00-35c0e68b0dac3c49be08a9d9cab32579-7ba0fc6d480c9841-01
Activity.ParentId:    00-35c0e68b0dac3c49be08a9d9cab32579-d95f666d24193f40-01
Activity.DisplayName: StepTwo
Activity.Kind:        Internal
Activity.StartTime:   2021-03-15T01:58:11.2894853Z
Activity.Duration:    00:00:01.0188689
Resource associated with Activity:
    service.name: unknown_service:temp

Activity.Id:          00-35c0e68b0dac3c49be08a9d9cab32579-d95f666d24193f40-01
Activity.DisplayName: SomeWork
Activity.Kind:        Internal
Activity.StartTime:   2021-03-15T01:58:10.7647839Z
Activity.Duration:    00:00:01.5450288
Resource associated with Activity:
    service.name: unknown_service:temp

Example work done
```

The console exporter is helpful for quick examples or local development but in a production deployment
you will probably want to send logs to a central logging store. OpenTelemetry supports a variety

> of different logging destinations using different

Reviewer 4: logging - I'll try to avoid this term :)

Author: Reworded : )

[exporters](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#exporter-library).
See the [OpenTelemetry getting started](https://github.com/open-telemetry/opentelemetry-dotnet#getting-started)
instructions for more information on configuring OpenTelemetry.

## Collect using Application Insights

Distributed tracing telemetry is automatically captured after configuring the Application Insights SDK 

> ([.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [.NET Core](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core))

Reviewer 4: ```suggestion
([.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [.NET Core](https://docs.microsoft.com/azure/azure-monitor/app/asp-net-core))
```

Reviewer 4: Need to do this for all doc links.

> ([.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [.NET Core](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core))

Reviewer 4: ```suggestion
([ASP.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [ASP.NET Core](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core))
```

or by enabling [code-less instrumentation](https://docs.microsoft.com/en-us/azure/azure-monitor/app/codeless-overview).

See the [Application Insights distributed tracing documentation](https://docs.microsoft.com/en-us/azure/azure-monitor/app/distributed-tracing) for more
information.

> [!NOTE]
> Currently Application Insights only supports collecting specific well-known Activity instrumentation and will ignore new user added Activities. Application
> Insights offers [TrackDependency](https://docs.microsoft.com/en-us/azure/azure-monitor/app/api-custom-events-metrics#trackdependency) as a vendor
> specific API for adding custom distributed tracing information.

> ## Collect using a custom logging implementation

Reviewer 5: Could we not use the word logging here? I was confused for a second on why we're talking about logging here 😄

Author: Reworded : )

        }

        static async Task DoSomeWork()
        {
            using (Activity a = s_source.StartActivity("SomeWork"))
            {
                await StepOne();
                await StepTwo();
            }
        }

        static async Task StepOne()
        {
            using (Activity a = s_source.StartActivity("StepOne"))
            {
                await Task.Delay(500);
            }
        }

        static async Task StepTwo()
        {
            using (Activity a = s_source.StartActivity("StepTwo"))
            {
                await Task.Delay(1000);
            }
        }
    }
}
```

Running the app does not log anything yet
```dotnetcli
> dotnet run
Example work done
```

### Add code to log the Activities

Update Main() with this code that logs Activities:
```C#
        static async Task Main(string[] args)
        {
            Activity.DefaultIdFormat = ActivityIdFormat.W3C;
            Console.WriteLine("         {0,-15} {1,-60} {2,-15}", "OperationName", "Id", "Duration");
            ActivitySource.AddActivityListener(new ActivityListener()
            {
                ShouldListenTo = (source) => true,

>                 Sample = (ref ActivityCreationOptions<ActivityContext> options) => ActivitySamplingResult.AllDataAndRecorded,

Reviewer 2: >Sample [](start = 16, length = 6)

did you try running this sample on Full Framework?

Author: Thanks, that did catch an error, there was an errant using System.Net.Http; left over that is harmless of Core but fails to build on net48.

                ActivityStarted = activity => Console.WriteLine("Started: {0,-15} {1,-60}", activity.OperationName, activity.Id),
                ActivityStopped = activity => Console.WriteLine("Stopped: {0,-15} {1,-60} {2,-15}", activity.OperationName, activity.Id, activity.Duration)
            });
            
            await DoSomeWork();
            Console.WriteLine("Example work done");
        }
```

The output now includes logging:
```dotnetcli
> dotnet run
         OperationName   Id                                                           Duration
Started: SomeWork        00-bdb5faffc2fc1548b6ba49a31c4a0ae0-c447fb302059784f-01
Started: StepOne         00-bdb5faffc2fc1548b6ba49a31c4a0ae0-a7c77a4e9a02dc4a-01
Stopped: StepOne         00-bdb5faffc2fc1548b6ba49a31c4a0ae0-a7c77a4e9a02dc4a-01      00:00:00.5093849
Started: StepTwo         00-bdb5faffc2fc1548b6ba49a31c4a0ae0-9210ad536cae9e4e-01
Stopped: StepTwo         00-bdb5faffc2fc1548b6ba49a31c4a0ae0-9210ad536cae9e4e-01      00:00:01.0111847
Stopped: SomeWork        00-bdb5faffc2fc1548b6ba49a31c4a0ae0-c447fb302059784f-01      00:00:01.5236391
Example work done
```

> Setting <xref:System.Diagnostics.Activity.DefaultIdFormat?displayProperty=nameWithType> is optional

Reviewer 2: >DefaultIdFormat [](start = 42, length = 15)

Should mention ForceDefaultIdFormat property too? 

I would suggest adding the usage of these properties to the sample with nice comment there.

Author: I added ForceDefaultIdFormat here and more discussion of both properties to the Conceptual doc which is linked.

> Setting <xref:System.Diagnostics.Activity.DefaultIdFormat?displayProperty=nameWithType> is optional

Reviewer 5: nit: This xref is broken

but helps ensure the sample produces similar output on different .NET runtime versions. .NET 5.0 uses

> the W3C ID format by default but earlier .NET versions default to using 

Reviewer 3: W3C -> W3C TraceContext

<xref:System.Diagnostics.ActivityIdFormat.Hierarchical?displayProperty=nameWithType> as a precaution
to avoid compatibility issues with older distributed tracing systems. See
[Activity IDs](distributed-tracing-concepts.md#activity-ids) for more details.

<xref:System.Diagnostics.ActivityListener?displayProperty=nameWithType> is used to receive callbacks
during the lifetime of an Activity. 
 - <xref:System.Diagnostics.ActivityListener.ShouldListenTo?displayProperty=nameWithType> - Each
Activity is associated with an ActivitySource which acts as a namespace for a set of Activities.
This callback is invoked once for each ActivitySource in the process. Returning true indicates
the listener should be notified about Activities associated with this source.

>  - <xref:System.Diagnostics.ActivityListener.Sample?displayProperty=nameWithType> - By default

Reviewer 5: nit: Broken xref <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType>

> <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType> does not

Reviewer 6: ```suggestion
<xref:System.Diagnostics.ActivitySource.StartActivity%2A?displayProperty=nameWithType> does not
```

This links to the overloads list for StartActivity. Is that what you want?

Author: I think the overload list would be fine, the statement is accurate for all overloads of the method. Thanks!

create an Activity object unless some ActivityListener indicates it should be sampled. Returning
<xref:System.Diagnostics.ActivitySamplingResult.AllDataAndRecorded?displayProperty=nameWithType>
indicates that the Activity should be created,
<xref:System.Diagnostics.Activity.IsAllDataRequested?displayProperty=nameWithType> should be set
to true, and <xref:System.Diagnostics.Activity.ActivityTraceFlags?displayProperty=nameWithType>
will have the <xref:System.Diagnostics.ActivityTraceFlags.Recorded?displayProperty=nameWithType>
flag set. IsAllDataRequested can be observed by the instrumented code as a hint that a listener
wants to ensure that auxilliary Activity information such as Tags and Events are populated.
The Recorded flag is encoded in the W3C ID and is a hint to other processes involved in the
distributed trace that this trace should be logged.
 - <xref:System.Diagnostics.ActivityListener.ActivityStarted?displayProperty=nameWithType> and
<xref:System.Diagnostics.ActivityListener.ActivityStopped?displayProperty=nameWithType> are
called when an Activity is started and stopped respectively. These callbacks provide an

> oportunity to log any relevant information about the Activity. When an Activity has just

Reviewer 4: ```suggestion
oportunity to add more information (Tags, Events, etc.) to the Activity. When an Activity has just
```

Author: Reworded a bit. I want to keep the emphasis on reading from the Activity but I did mention that it is possible to modify it as well.

started much of the data may still be incomplete and it will be filled in before the Activity
stops.

Once an ActivityListener has been created and the callbacks are populated, invoking

> <xref:System.Diagnostics.ActivitySource.AddActivityListener?displayProperty=nameWithType>

Reviewer 6: ```suggestion
<xref:System.Diagnostics.ActivitySource.AddActivityListener(System.Diagnostics.ActivityListener)?displayProperty=nameWithType>
```

---
title: Collect a distributed trace - .NET
description: Tutorials to collect distributed traces in .NET applications using OpenTelemetry, Application Insights, or ActivityListener
ms.topic: tutorial
ms.date: 03/14/2021
---

# Collect a distributed trace

**This article applies to: ✔️** .NET Core 2.1 and later versions **and** .NET Framework 4.5 and later versions

Instrumented code can create <xref:System.Diagnostics.Activity> objects as part of a distributed trace, but the information
in these objects needs to be collected into a centralized persistant store so that the entire trace can be
reviewed later. In this tutorial you will collect the distributed trace telemetry in different ways so that it is
available to diagnose application issues when needed. See
[the instrumentation tutorial](distributed-tracing-instrumentation-walkthroughs.md) if you need to add new instrumentation.

## Collect traces using OpenTelemetry

### Prerequisites

> - [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet) or a later version

Reviewer 4: nit: to avoid the misconception that "Collect traces using OpenTelemetry" *requires* .NET Core 3.1, I'd suggest to keep this .NET Core 2.1

version 5 or greater.

```dotnetcli
dotnet add package System.Diagnostics.DiagnosticSource
```

Replace the contents of the generated Program.cs with this example source:

```C#
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace Sample.DistributedTracing
{
    class Program
    {
        static ActivitySource s_source = new ActivitySource("Sample.DistributedTracing");

        static async Task Main(string[] args)
        {
            await DoSomeWork();
            Console.WriteLine("Example work done");
        }

        static async Task DoSomeWork()
        {
            using (Activity a = s_source.StartActivity("SomeWork"))
            {
                await StepOne();
                await StepTwo();
            }
        }

        static async Task StepOne()
        {
            using (Activity a = s_source.StartActivity("StepOne"))
            {
                await Task.Delay(500);
            }
        }

        static async Task StepTwo()
        {
            using (Activity a = s_source.StartActivity("StepTwo"))
            {
                await Task.Delay(1000);
            }
        }
    }
}
```

> Running the app does not record any tracing information yet

Reviewer 4: ```suggestion
Running the app does not collect any trace data yet:
```

```dotnetcli
> dotnet run
Example work done
```

### Collect using OpenTelemetry

[OpenTelemetry](https://opentelemetry.io/) is a vendor neutral open source project supported by the
[Cloud Native Computing Foundation](https://www.cncf.io/) that aims to standardize generating and collecting telemetry for
cloud-native software. In this example you will collect and display distributed trace information on the console though
OpenTelemetry can be reconfigured to send it elsewhere. See the
[OpenTelemetry getting started guide](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started/README.md)
for more information.

> Add the [OpenTelemetry](https://www.nuget.org/packages/OpenTelemetry/) and

Reviewer 4: Installing `OpenTelemetry.Exporter.Console` automatically brings `OpenTelemetry`, so this step may be implict.

```

Replace the contents of the generated Program.cs with this example source:

```C#
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace Sample.DistributedTracing
{
    class Program
    {
        static ActivitySource s_source = new ActivitySource("Sample.DistributedTracing");

        static async Task Main(string[] args)
        {
            await DoSomeWork();
            Console.WriteLine("Example work done");
        }

        static async Task DoSomeWork()
        {
            using (Activity a = s_source.StartActivity("SomeWork"))
            {
                await StepOne();
                await StepTwo();
            }
        }

        static async Task StepOne()
        {
            using (Activity a = s_source.StartActivity("StepOne"))
            {
                await Task.Delay(500);
            }
        }

        static async Task StepTwo()
        {
            using (Activity a = s_source.StartActivity("StepTwo"))
            {
                await Task.Delay(1000);
            }
        }
    }
}
```

Running the app does not collect any trace data yet:

```dotnetcli
> dotnet run
Example work done
```

> ### Add code to observe the Activities

Reviewer 4: ```suggestion
### Add code to collect the traces
```

Started: StepTwo         00-bdb5faffc2fc1548b6ba49a31c4a0ae0-9210ad536cae9e4e-01
Stopped: StepTwo         00-bdb5faffc2fc1548b6ba49a31c4a0ae0-9210ad536cae9e4e-01      00:00:01.0111847
Stopped: SomeWork        00-bdb5faffc2fc1548b6ba49a31c4a0ae0-c447fb302059784f-01      00:00:01.5236391
Example work done
```

Setting <xref:System.Diagnostics.Activity.DefaultIdFormat> and
<xref:System.Diagnostics.Activity.ForceDefaultIdFormat> is optional
but helps ensure the sample produces similar output on different .NET runtime versions. .NET 5 uses
the W3C TraceContext ID format by default but earlier .NET versions default to using
<xref:System.Diagnostics.ActivityIdFormat.Hierarchical> ID format. See
[Activity IDs](distributed-tracing-concepts.md#activity-ids) for more details.

<xref:System.Diagnostics.ActivityListener?displayProperty=nameWithType> is used to receive callbacks
during the lifetime of an Activity.

- <xref:System.Diagnostics.ActivityListener.ShouldListenTo> - Each
Activity is associated with an ActivitySource which acts as a namespace for a set of Activities.
This callback is invoked once for each ActivitySource in the process. Returning true indicates

> the listener should be notified about Activities associated with this source.

Reviewer 4: you may add something to the tune of "Returning false means the `Sample` callback is not invoked at all.."

Author: I updated the text for 'true' path to make it clear you needed to pick that if you wanted to do sampling. Hopefully that is explicit enough but we can always make further edits as needed.
